### PR TITLE
MAINT: unify factorial implementations

### PR DESCRIPTION
**Pull request by h-vetinari**
After a bunch of work on the factorial functions (in terms of unifying API and code paths), we should now be able to switch to a single unified implementation without any extra cost. Despite needing some minor indirection (e.g. to distinguish warning messages), I believe it's vastly preferable to have one implementation, rather than 3 that are very similar.

Follows (i.a.):
* #18409
* #19716
* #19989
* #21702
* #21801

### CODE REVIEW

File: scipy/special/_basic.py [hunk 1]
```diff
@@ -3182,63 +3229,10 @@ def factorial2(n, exact=False, extend="zero"):
...
-    if n.size == 0:
-        # return empty arrays unchanged
-        return n
-    elif exact:
-        return _factorialx_array_exact(n, k=2)
-    return _factorialx_array_approx(n, k=2, extend=extend)
-
-
-def factorialk(n, k, exact=None, extend="zero"):
+def factorialk(n, k, exact=False, extend="zero"):
```
Comment (j-bowhay) [inline scipy/special/_basic.py:3235]: Here and elsewhere should we make the new input keyword only?

```suggestion
def factorialk(n, k, exact=False, *, extend="zero"):
```

Comment (h-vetinari) [reply scipy/special/_basic.py:3235]: I thought about it - not sure. If we do it should IMO be at the [right](https://github.com/scipy/scipy/issues/18703#issuecomment-1659814261) place, not just for the new keyword. I even looked up my mailing list [post](https://mail.python.org/archives/list/scipy-dev@python.org/thread/MK7AWN5ARX7AOJXDANUPTFYHSJ5Y53ES/) on this (already >1y 😱). I guess we'd be ready for a large-scale change-over, as soon as people are comfortable with discussing SciPy 2.0...

Comment (j-bowhay) [reply scipy/special/_basic.py:3235]: Fair point

**Review by j-bowhay — APPROVED** [review]
Very nice to see this coming together

**Pull request merged**